README.md

```diff
@@ -0,0 +1,81 @@
+# Taskcluster Component Loader
+
+A taskcluster component loader returns a function that can be used to
+selectively initialize components.  The main impetus is that we have a lot
+of boiler plate that is duplicated in different `bin/*.js` files and in
+`test/helper.js` files around initializing components.  Rather than maintain
+this boiler plate code, we've decided to allow an implementer to write out
+how their components are initialized once and use the same initialization logic in all
+of their scripts and testing.
+
+One of the design goals was to be able to have separate loaders in the same
+process.  This makes it easier to initialize the same component with a
+different configuration for testing.
+
+There are two layers of 'component directories'.  This is so that we can
+have a single static file which explains how things like Entities,
+Validators, Publishers and APIs are created but that don't have the
+information about things like configuration values.  In this file, we will
+have components that depend on components that doesn't exist in the file.
+This is solved by specifying a set of 'virtual components' by name which are
+those components that will be described by the second level of component
+directory.  These 'virtual dependencies' are things like how to read a
+configuration file/format and provide the values that the components in the
+static file know how to consume.  We ensure that the list of required dependencies
+is specified to make error checking easier.
+
+All values returned by the loader are promises.  If you wish to use the
+value of something from the loader, you must await the dependency value from
+an async function.  If you're using simple promises, just .then() the value
+in your code.
+
+A component is either an object with a setup method or a flat value.  Flat
+values are just values wrapped in a Promise.resolve() call to ensure a
+consistent API.  Objects with a setup method are required and able to
+specify dependencies.  This object is in the form:
+```javascript
+loader({
+  name: {
+    requires: ['A'],
+    setup: async deps => {
+       console.log(await deps.A);
```

Comment:
Why would consuming code need to know that.... See code snippets I added here they do it..